gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java

```diff
@@ -0,0 +1,335 @@
+          public ChangeRequest apply(com.google.api.services.dns.model.Change changePb) {
+            return fromPb(changePb);
+          }
+        };
+    private static final long serialVersionUID = -8737501076674042014L;
+    private final String zoneName;
+    private final Map<DnsRpc.Option, ?> requestOptions;
+    private final DnsOptions serviceOptions;
+
+    ChangeRequestPageFetcher(String zoneName, DnsOptions serviceOptions, String cursor,
+        Map<DnsRpc.Option, ?> optionMap) {
+      this.zoneName = zoneName;
+      this.requestOptions =
+          PageImpl.nextRequestOptions(DnsRpc.Option.PAGE_TOKEN, cursor, optionMap);
+      this.serviceOptions = serviceOptions;
+    }
+
+    @Override
+    public Page<ChangeRequest> nextPage() {
+      return listChangeRequests(zoneName, serviceOptions, requestOptions, PB_TO_CHANGE_REQUEST);
+    }
+  }
+
+  private static class DnsRecordPageFetcher implements PageImpl.NextPageFetcher<DnsRecord> {
+
+    private static final Function<ResourceRecordSet, DnsRecord> PB_TO_DNS_RECORD =
+        new Function<ResourceRecordSet, DnsRecord>() {
+          @Override
+          public DnsRecord apply(com.google.api.services.dns.model.ResourceRecordSet pb) {
+            return DnsRecord.fromPb(pb);
+          }
+        };
+    private static final long serialVersionUID = 670996349097667660L;
+    private final Map<DnsRpc.Option, ?> requestOptions;
+    private final DnsOptions serviceOptions;
+    private final String zoneName;
+
+    DnsRecordPageFetcher(String zoneName, DnsOptions serviceOptions, String cursor,
+        Map<DnsRpc.Option, ?> optionMap) {
+      this.zoneName = zoneName;
+      this.requestOptions =
+          PageImpl.nextRequestOptions(DnsRpc.Option.PAGE_TOKEN, cursor, optionMap);
+      this.serviceOptions = serviceOptions;
+    }
+
+    @Override
+    public Page<DnsRecord> nextPage() {
+      return listDnsRecords(zoneName, serviceOptions, requestOptions, PB_TO_DNS_RECORD);
+    }
+  }
+
+  @Override
+  public Page<Zone> listZones(ZoneListOption... options) {
+    return listZones(options(), optionMap(options));
+  }
+
+  private static Page<Zone> listZones(final DnsOptions serviceOptions,
+      final Map<DnsRpc.Option, ?> optionsMap) {
+    // define transformation function
+    // this differs from the other list operations since zone is functional and requires dns service
+    Function<ManagedZone, Zone> pbToZoneFunction = new Function<ManagedZone, Zone>() {
+      @Override
+      public Zone apply(
+          com.google.api.services.dns.model.ManagedZone zonePb) {
+        return new Zone(serviceOptions.service(), ZoneInfo.fromPb(zonePb));
+      }
+    };
+    try {
+      // get a list of managed zones
+      DnsRpc.ListResult<ManagedZone> result =
+          runWithRetries(new Callable<DnsRpc.ListResult<ManagedZone>>() {
+            @Override
+            public DnsRpc.ListResult<ManagedZone> call() {
+              return serviceOptions.rpc().listZones(optionsMap);
```

Comment:
As per discussion, kept consistent. However, pulled rpc() call out of the retries using a local variable.